dashboard,syz-ci: implement asset storage

a-nogikh opened the pull request:
Asset is a file/attachment relevant for debugging a kernel crash or the
syzbot itself.

**TODO list:**

- <s>Fix lint errors and unit tests</s>
- <s>Deploy and debug a test instance.</s>
- <s>Generate the new `index.yaml` file.</s>
- Write some documentation.

**Dashboard** keeps track of the uploaded assets and manages their lifetime.
Some of the assets are attached to the bug reports sent by syzkaller,
some (like html coverage report) are displayed on the web portal.

Two new dashboard API requests:
* add_build_asset -- let dashboard remember one more build-related
  asset.
* needed_assets -- query the list of assets that are still needed by the
  dashboard.

**The syz-ci side** of the asset storage functionality is responsible for the following:
* Intercept assets at various stages of syz-ci operation.
* Compress and upload assets to GCS.
* Report assets to the dashboard.
* Remove no longer needed assets.

a-nogikh commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,577 @@
+*/
+
+/*
+   Alternative DB design: don't embed assets, but store them as separate entities.
```
Looking at how complex the deprecated asset detection gets when we embed Assets (see the implemented code), I'm leaning more and more towards this somewhat simpler approach. 

@dvyukov, what do you think?

a-nogikh replied:
FTR after an offline discussion it was decided to switch from querying assets to forget to querying assets to remember. This leads to fewer corner cases.

dvyukov commented on pkg/assets/config.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+type Config struct {
+	GcsBucket string                     `json:"gcs_bucket"`
```
This looks too GCS-specific. syz-ci can upload coverage reports to FTP. It uses a more generic "where to upload" name and them looks at the value prefix ("gs:", "ftp:") to choose where to upload.
It's better to use the same here.

dvyukov reviewed (commented):
Some naming comments that should be easier to address while you still have a WIP version in a single patch.

a-nogikh commented on pkg/assets/storage.go:
```diff
@@ -0,0 +1,362 @@
+		return nil, err
+	}
+	if req.contentType != "" || req.contentEncoding != "" {
+		err = csb.client.SetContentMetadata(path, req.contentType, req.contentEncoding)
```
> It feels that this needs to be a part of UploadFile itself.

You mean `UploadFile` from `pkg/gce/gce.go`, right?

dvyukov replied:
The one from pkg/gcs/gcs.go.

a-nogikh commented on pkg/assets/types.go:
```diff
@@ -0,0 +1,30 @@
+	BootableDisk       = "bootable_disk"
+	NonBootableDisk    = "non_bootable_disk"
+	KernelObject       = "kernel_object"
+	VmLinux            = "vmlinux"
```
Do we really need to make these types as generic as possible? `vmlinux` anyway looks like a totally different file type from files that play the same role for other OSes, so from the user perspective it'd be more clear if they were also designated differently in an email. I.e. `vmlinux file: %URL%` looks better than `kernel debug object: %URL%`.

dvyukov replied:
> Do we really need to make these types as generic as possible?

It depends on what are the reasons to make it non-generic. If there are no major reasons to make code non-generic, it's always better to have it generic from multiple perspectives.

Mapping them to different user-readable strings for different OSes does not look like a sufficient reason to make code non-generic on the protocol level.

We already have a notion of "kernel object" and can be different files for different OSes:
https://github.com/google/syzkaller/blob/1434eec0b84075b7246560cfa89f20cdb3d8077f/sys/targets/targets.go#L75-L76

These are not really totally different file types, these are all ELF files. And we have generic code that runs nm/objdump on them. If we save raw coverage, I assume we may want to have the same common handling for the kernel object files stored in the asset storage.

You already have the function that maps these types to user-readable strings, it would be reasonable to make that function to accept OS/target and produce different user-readable strings if that's what we want. Btw, maybe it could use the sys/targets.KernelObject names for this since these should be familiar to kernel developers.

a-nogikh replied:
Thanks for the clarification, fixed.

dvyukov commented on dashboard/dashapi/dashapi.go:
```diff
@@ -592,6 +604,15 @@ func (dash *Dashboard) Query(method string, req, reply interface{}) error {
 		dash.logger("API(%v): %#v", method, req)
 	}
+	if dash.mocker != nil {
+		if reply == nil {
```
I think we can replace this with just:
`return dash.mocker(method, req, resp)`

a-nogikh commented:
Merged the helper commits as a separate PR and rebased this one.

dvyukov commented on syz-ci/syz-ci.go:
```diff
@@ -265,6 +271,34 @@ func main() {
+	}
+	storage, err := asset.StorageFromConfig(cfg.AssetStorage, dash)
+	if err != nil {
+		log.Logf(0, "deprecateAssets: failed to create asset storage: %v", err)
```
Nobody is reading all logs every day, so just logging an error is not much different from silently ignoring it. Why not log.Fatalf?

dvyukov commented on syz-ci/syz-ci.go:
```diff
@@ -265,6 +271,34 @@ func main() {
+		log.Logf(0, "deprecateAssets: failed to create asset storage: %v", err)
+		return
+	}
+	rand.Seed(time.Now().UTC().UnixNano())
```
If we need to do this, this needs to be done in main. Otherwise different small functions everywhere will re-seed it and log different seeds, which will be confusing. This is a global things, so it needs to be setup in main.
Or better use a local rand instance.
But also why do we need this randomization here? Why not run it, say, once per day?

a-nogikh replied:
The reasoning was to avoid (almost) synchronous requests from a lot of syz-ci's to the dashboard.

Though now that deprecated assets querying was turned into the much faster needed assets querying, it's probably no longer so important indeed. I'll remove the randomization.

dvyukov commented on syz-ci/syz-ci.go:
```diff
@@ -265,6 +271,34 @@ func main() {
+			break loop
+		case <-time.After(time.Hour * time.Duration(sleepHours)):
+		}
+		log.Logf(0, "running deprecateAssets()")
```
In Go function names are not generally followed by (), it's not really necessary and does not add much info. That's typically done in C.
Also we tend to not include function names in log messages:
https://gist.githubusercontent.com/dvyukov/ad3957446f8a9239b4c1ad4f0575316b/raw/2161ea2e644b65659e2be42879ff6838c9e6b43b/gistfile1.txt
but rather say something more in user terms (e.g. "deprecating assets...").

dvyukov commented on syz-ci/syz-ci.go:
```diff
@@ -256,6 +261,7 @@ func main() {
 		}
 	})
 
+	go deprecateAssets(cfg, shutdownPending)
```
I think it's better to also do wg.Add/Done for deprecateAssets to not abort it in the middle.

dvyukov commented on syz-ci/manager.go:
```diff
@@ -99,7 +103,13 @@ func createManager(cfg *Config, mgrcfg *ManagerConfig, stop chan struct{}, debug
+	if !cfg.AssetStorage.IsEmpty() {
+		assetStorage, err = asset.StorageFromConfig(cfg.AssetStorage, dash)
+		if err != nil {
+			log.Logf(0, "failed to create asset storage: %v", err)
```
Same here - nobody is reading the logs.

dvyukov commented on syz-ci/syz-ci.go:
```diff
@@ -265,6 +271,34 @@ func main() {
+		log.Logf(0, "running deprecateAssets()")
+		err := storage.DeprecateAssets()
+		if err != nil {
+			log.Logf(0, "deprecateAssets() failed: %v", err)
```
Use dash.LogError, see other uses in syz-ci.

dvyukov commented on syz-ci/manager.go:
```diff
@@ -62,27 +63,30 @@ func init() {
+	storage         *asset.Storage
+	stop            chan struct{}
+	debug           bool
+	lastBuildObject *dashapi.Build
```
Half of the fields in Manager are objects, appending "Object" does not add lots of useful information for the reader. "lastBuild" is a fine name.

dvyukov commented on syz-ci/manager.go:
```diff
@@ -634,7 +651,55 @@ func (mgr *Manager) pollCommits(buildCommit string) ([]string, []dashapi.Commit,
+}
+
 func (mgr *Manager) uploadCoverReport() error {
+	if mgr.storage == nil {
```
syz-ci may with w/o dashboard/asset storage and upload coverage reports to FTP.

dvyukov commented on syz-ci/manager.go:
```diff
@@ -634,7 +651,55 @@ func (mgr *Manager) pollCommits(buildCommit string) ([]string, []dashapi.Commit,
+		return fmt.Errorf("no asset storage is enabled")
+	}
+	if mgr.lastBuildObject == nil {
+		return fmt.Errorf("no build has been uploaded yet")
```
When can this happen?

dvyukov commented on syz-ci/manager.go:
```diff
@@ -634,7 +651,55 @@ func (mgr *Manager) pollCommits(buildCommit string) ([]string, []dashapi.Commit,
+		if !mgr.storage.AssetTypeEnabled(pendingAsset.assetType) {
+			continue
+		}
+		go mgr.storage.UploadBuildAssetCopy(pendingAsset.path,
```
What's the reason for using "Copy" version here?

dvyukov commented on syz-ci/manager.go:
```diff
@@ -555,6 +571,7 @@ func (mgr *Manager) uploadBuild(info *BuildInfo, imageDir string) (string, error
 	if err := mgr.dash.UploadBuild(build); err != nil {
 		return "", err
 	}
+	mgr.uploadBuildArtifacts(build, imageDir)
```
A simpler way to do this is to upload assets first, then upload the build info to the dashboard once with all of the assets included; rather than upload assets afterwards and add them one-by-one. This way the dashboard API is simpler and we don't have half-filled objects in the datastore.
What's the advantage of doing it the current way?

dvyukov replied:
Errr... cover reports :(

We may already report a bug (in particular for broken build) but don't have build artifacts yet.

Also we may upload the same build multiple times, e.g. LTS releases may change once per 1-2 weeks, but we restart and upload build every day. Currently we will re-upload duplicates of images to GCS, right? Could we avoid that?

a-nogikh replied:
Now the code only uploads build assets after a rebuild is done (unless we're in debug mode). Also, adjusted the API -- now it can take assets both per separate API call or during build creation.

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+}
+
+// TODO: set it in the config files.
+func DefaultPreprocessors() *PreprocessorCollection {
```
Enexport.

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+var PreprocessXz = Preprocessor{
+	Available: func() bool {
+		xzCheck.Do(func() {
+			_, err := osutil.RunCmd(time.Minute, "", "xz --version")
```
Have you considered using some Go package for this?
The tool needs to be installed separately, not sure if it's available on *BSD, Mac, Windows, Fuchsia. May have different interface there.
If we do this, it will also remove all of the Available-related code code.

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+	},
+}
+
+// TODO: set it in the config files.
```
How configurable do we need this to be?
Can we simply always do some good compression?
The contentType for html does not seem to be something that should be configured at all. We need the "text/html" contentType on the html report always. Using some other content type will break things. So it probably needs to be specified on upload request, or be a property of asset type (if it's an html coverage report, we already know what content type it needs).

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+			return nil, fmt.Errorf("gzip preprocess: failed to open the tmp file: %w", err)
+		}
+		defer file.Close()
+		writer := bufio.NewWriter(file)
```
Do we need this?

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+		if err != nil {
+			return nil, fmt.Errorf("gzip preprocess: NewWriterLevel failed: %w", err)
+		}
+		io.Copy(gzip, origFile)
```
Check for errors.

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+			return nil, fmt.Errorf("gzip preprocess: NewWriterLevel failed: %w", err)
+		}
+		io.Copy(gzip, origFile)
+		gzip.Close()
```
Check for errors.

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+		}
+		io.Copy(gzip, origFile)
+		gzip.Close()
+		writer.Flush()
```
Check for errors.

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+		}
+		newReq := *req
+		newReq.origFile = tmpFile.Name()
+		newReq.savePath = fmt.Sprintf("%s.xz", newReq.savePath)
```
Can we use contentEncoding as we do for gzip?

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+type PreprocessorCollection struct {
+	defaultOne Preprocessor
+	custom     map[Type]Preprocessor
+	// Map cannot be concurrently accessed.
```
We don't change it after creation so we don't need any protection. Reading from the map can be done concurrnetly.

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+	return storage.UploadBuildAsset(tmpFile, assetType, build)
+}
+
+func (storage *Storage) UploadBuildAssetCopy(origFile string, assetType Type, copyName string,
```
With 3 different versions UploadBuildAsset/UploadBuildAssetStream/UploadBuildAssetCopy it feels that the package it trying to be too helpful for callers. Only 2 of these versions are actually used. If also upload Crash, Bug and Manager artifacts, we will need 12 different permutations. What these variations add feels like osutil package territory. If we add something like osutil.WriteFileReader, it may be enough. Previously we avoided saving whole coverage report to disk and directly streamed it to GCS. Now we save it multiple times to disk anyway, so there is little point in providing special version for io.Reader.

a-nogikh replied:
Hmm, it actually feels like moving everything to the streams might be the most generic option. Gzip compression/decompression is built on streams, xz can also operate on stdin/stdout (`echo "abd" | xz -0 -c | xxd -p`). So we can possibly get rid of all copying during internal processing.

But the problem is that sometimes we'd prefer to still copy a file before uploading it. E.g. for the case when uploading turned out to be very slow and the underlying file might have already been replaced by the newly compiled kernel. This is very unrealistic on syzbot, but we still need to somehow take that into account in the code.

Though that copying might be actually offloaded to `osutil` (e.g. `func CopyFileCallback(sourceFile string, cb func(tmpFile string)error) error`).

WDYT?

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+	period time.Duration) (map[string]Asset, error) {
+	// We don't want to query everything for such a purpose.
+	// Assume the assets we're interested in are uploaded often enough.
+	const queryLastAssetsFor = time.Hour * 24 * 14
```
This needs to use period, or don't accept period.

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+	_, err := db.NewQuery("Build").
+		Filter("Assets.Type=", assetType).
+		Filter("Assets.CreateDate>", startTime).
+		Order("Assets.CreateDate").
```
Needs to use ns.

a-nogikh commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+	fileExt := filepath.Ext(baseName)
+	name := fmt.Sprintf("%s-%s%s",
+		strings.TrimSuffix(baseName, fileExt),
+		build.KernelCommit,
```
This was meant to maybe help the user to distinguish various downloaded assets, and kernel commit hash looked like a good anchor point. In GCS these files will have differing names anyway, as we later prepend a "folder" to the file path with the current time.

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+	return nil
+}
+
+const buildBatchSize = 16
```
move inside of the function, just before the use
it's used once and the name is generic

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+			errors <- callFunc()
+		}(f)
+	}
+	defer close(errors)
```
Remove, not necessary.

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+		errorStr = fmt.Sprintf("%s%v\n", errorStr, err)
+	}
+	if errorStr != "" {
+		return fmt.Errorf("%s\n%s", errorStart, errorStr)
```
I feels that this `errorStart` does not belong to this pure abstract function.
Each caller does the following after calling parallelize anyway:
```
if err != nil {
  return err
}
```
and this looks like the right place to wrap the error with any additional context if necessary, or maybe it's not necessary, or maybe the context must be not just a string, but say also include namespace or something else.

Potentially we can then build error type directly and have just `return err` after the loop... but not sure if it will be better than using the string.

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+//    we need to handle extra corner cases:
+//    - If JobDoneReq failed, assets must be deleted.
+//    - If we uploaded several files, all must be deleted.
+//    - What if syz-ci was killed somewhere in between? How do we get rid of those abadoned files?
```
Are all of the cases automatically handled by the current algorithm?
We will see these abandoned assets are not needed anymore and automatically delete them after some time with no additional code, no?

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+		return nil
+	}
+	if err := db.RunInTransaction(c, tx, &db.TransactionOptions{}); err != nil {
+		log.Errorf(c, "failed to update build: %v", err)
```
Shouldn't any request errors be already logged somewhere up the stack?
We don't generally log whenever we return an error.
Then this can be just `return db.RunInTransaction(...)`

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+		if err != nil {
+			return err
+		}
+		build.Assets = append(build.Assets, asset)
```
Need to somehow deduplicate by asset type.

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+	return nil
+}
+
+func handleDeprecateAssets(w http.ResponseWriter, r *http.Request) {
```
Please revert order of all functions involved in asset deprecation. It's very inconvenient to read text bottom-up. handleDeprecateAssets should be the first at the top.

dvyukov commented on dashboard/app/main.go:
```diff
@@ -1126,12 +1142,21 @@ func loadManagers(c context.Context, accessLevel AccessLevel, ns, manager string
 		if now.Sub(mgr.LastAlive) > 6*time.Hour {
 			uptime = 0
 		}
+		// TODO: ideally we'd want to explicitly detect and mark cases, when
```
What would we do differently for such builds?

a-nogikh replied:
We'd add some note that attached the coverage report is related to the previous build, not to the current one. It should help the user understand what exactly they are downloading.

dvyukov replied:
I am not sure users care about current build/previous build. They may want to know the kernel commit, and I assume we could show it somehow for the report (but nobody asked for this).

a-nogikh replied:
Users in general probably don't care, but I'd prefer to see if the coverage report is (already) worth looking at. Adjusted the comment.

dvyukov commented on dashboard/app/main.go:
```diff
@@ -1126,12 +1142,21 @@ func loadManagers(c context.Context, accessLevel AccessLevel, ns, manager string
+		// the build is too new to have some coverage report. For now we just
+		// query the latest uploaded cover report for the manager.
+		coverURL := ""
+		if config.CoverPath != "" {
```
How will we configure this? syz-ci's won't upload reports using the old scheme anymore, right? So does it make sense to have CoverPath set for any kernels?

a-nogikh replied:
Adjusted the code -- now syz-ci accepts and handles `CoverPath` if no asset storage is configured

dvyukov commented on dashboard/dashapi/dashapi.go:
```diff
@@ -537,6 +546,33 @@ func (dash *Dashboard) UploadManagerStats(req *ManagerStatsReq) error {
+// Asset lifetime:
+// 1. syz-ci uploads it to GCS and reports to the dashboard via add_build_asset.
+// 2. syz-ci periodically queries deprecated_assets to figure out which assets
+//
```
remove empty line

dvyukov commented on dashboard/dashapi/dashapi.go:
```diff
@@ -537,6 +546,33 @@ func (dash *Dashboard) UploadManagerStats(req *ManagerStatsReq) error {
+//
+//	can be deleted.
+//
+// 3. Once an asset is deleted, syz-ci invokes forget_assets.
```
This is the only place that mentions forget_assets.

dvyukov commented on pkg/asset/storage.go:
```diff
@@ -0,0 +1,369 @@
+
+var ErrAssetTypeDisabled = errors.New("uploading assets of this type is disabled")
+
+func (storage *Storage) UploadFile(origFile string, assetType Type, name string) (string, error) {
```
Users are intended to use UploadBuildAsset*, right? If so,  unexport this.

a-nogikh commented on pkg/asset/backend_dummy.go:
```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
```
It actually turned out easier to rename the type itself to match the name. Otherwise we have problems testing the object created via `StorageFromConfig` function.

a-nogikh commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+	ret := map[string]Asset{}
+	for _, build := range builds {
+		for _, asset := range build.Assets {
+			if asset.Type != assetType {
```
That filter would only query builds that have at least one asset of the necessary type and create date. If at least one asset matches, the returned build object will include everything that's in the DB for the object -- i.e. also those assets that don't match these filters.

a-nogikh commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,347 @@
+	var builds []*Build
+	_, err := db.NewQuery("Build").
+		Filter("Assets.DownloadURL>", "").
+		GetAll(c, &builds)
```
I thought about that, but it looks like it's just not worth it.

Why not per-namespace:
1) At the syz-ci level, we don't really know which namespace each syz-manager belongs. In the best case, we could group them by dashboard client and deprecate assets separately, but that looks somewhat excessive.
2) The current algorithm requires us to subtract the set of the existing assets and the needed ones. It means, we'd have to also store the assets separately, so that we would be able to only query GCS objects belonging to some specific namespace (and again -- we don't even know the namespace of each instance). Otherwise we'd just quickly end up with 0 remaining assets.
3) Once we implement crash-specific assets, it will be painful to query the assets from crashes, which are related to some specific namespace.

Storing them per-namespace looks a bit better -- at least we know the names of the managers, but (2) holds -- we'd need to also store them separately.

Another question is what performance benefits it can bring. If all syz-ci instances share the same GCS bucket, asset deprecation can be delegated to only one of them, so we won't really make a lot of unnecessary querying. And even if each instance queries them, it doesn't seem that we are going to have really a lot of assets.

We can probably prepend the manager name prefix before GCS object names (as "folders"), so that we could later query them per-manager, should it turn necessary. GCS API provides the ability to filter objects by path prefix: https://pkg.go.dev/cloud.google.com/go/storage#Query

WDYT?

a-nogikh commented on dashboard/app/asset_storage_test.go:
```diff
@@ -0,0 +1,242 @@
+Downloadable assets:
+
+vmlinux:
+http://google.com/vmlinux
```
Yes, I considered them, but I fear that the lines would be just too incomprehensible. Your example looks short and nice, but let's look at what there'll actually be:

```
Downloadable assets:
vmlinux: https://storage.googleapis.com/syzkaller-assets/1658254186063479634/vmlinux-4a57a8400075.xz
disk image (non-bootable): https://storage.googleapis.com/syzkaller-assets/1658254186064632511/non_bootable_disk-4a57a8400075.raw
```

Will it be ok to insert such long lines to the emails?

dvyukov replied:
Hmm... these links don't look nice... thinking how we could shorten them.
1. we could shorten these varying parts "1658254186063479634" and "4a57a8400075".
2. use only 1 varying part "1658254186063479634" or "4a57a8400075".
3. use some link shortener for the common prefix "https://storage.googleapis.com/syzkaller-assets".
4. use some link shortener for the whole link.
5. shorten them ourselves and give something like https://syzkaller.appspot.com/a/108f654fb4a57a84 (this is only 48 chars).

a-nogikh replied:
I made the links somewhat shorter. Link shorteners are a nice idea, but I'd prefer to implement it separately, not as a part of this PR.

Now the PR would generate reports like that
```
Downloadable assets:
vmlinux: https://storage.googleapis.com/syz-assets/3e3e8a9ed730/vmlinux-4a57a840.xz
disk image (non-bootable): https://storage.googleapis.com/syz-assets/4d65d99632f6/non_bootable_disk-4a57a840.raw.xz
```

dvyukov commented on dashboard/app/asset_storage.go:
```diff
@@ -0,0 +1,368 @@
+	return nil
+}
+
+func parallelize(errorStart string, funcs ...func() error) error {
```
`errgroup.Group` may be an existing altearntive:
https://www.storj.io/blog/production-concurrency
https://pkg.go.dev/golang.org/x/sync/errgroup

a-nogikh replied:
Indeed! Fixed.

a-nogikh commented on dashboard/app/reporting.go:
```diff
@@ -366,6 +367,7 @@ func reproStr(level dashapi.ReproLevel) string {
 	}
 }
 
+// nolint: gocyclo
```
I think it's better to do it as a separate PR, this one is already too big.

a-nogikh commented:
@dvyukov PTAL

I addressed your comments (for the non-implemented ones see my replies).

There's a [all: code review fixes](https://github.com/google/syzkaller/pull/3223/commits/f886f4fc03402a5eda5f9f0e9da3b1bb40d86a55) commit that is still there only for convenience purposes. I'll split and squash it when we get to merging this PR.

dvyukov commented on dashboard/dashapi/dashapi.go:
```diff
@@ -402,11 +403,21 @@ type BugReport struct {
+// Asset types used throughout the system.
+// DO NOT change them, this will break compatibility with DB content.
+const (
+	BootableDisk       AssetType = "bootable_disk"
```
This is of type string only for syz-ci json config, right?

If so, I would switch this to integer codes and add `asset.TypeDescription.configName` for the config.
Then per-type configs can be retrieved with `c.Assets[assetTypes[assetType].configName]`.

a-nogikh replied:
This also makes it easier to e.g. debug-print data structures and to manually explore the DB/raw dashboard requests. Are there substantial reasons to prefer integers in this case?

dvyukov replied:
Strings are for humans. Numbers are for machines.
These are stored both in DB and in the config. This means we can never ever rename config names. And people don't always read comments that something shouldn't be renamed.

If you want debug printing add `String() string` method. fmt/log will use it.
And we will be able to reuse this method to access the config as `c.Assets[assetType.String()]`.

a-nogikh replied:
> And people don't always read comments that something shouldn't be renamed.

One more risk I see is that we become dependent on the exact order of asset type consts (in case of `iota`) or take extra care when assigning int values ourselves. So it's tricky either way.

If the biggest problem is in renaming config-file-facing consts, maybe we could still save strings in the DB, and (once it's needed) just add something like `customConfigName` to `TypeDescription` in `pkg/asset/type.go` and adjust `pkg/asset/config.go` to be able to understand it?

a-nogikh replied:
// During offline discussion it was decided to leave this as is.

dvyukov commented on syz-ci/manager.go:
```diff
@@ -571,7 +582,24 @@ func (mgr *Manager) uploadBuild(info *BuildInfo, imageDir string) (string, error
+				mgr.Errorf("failed to upload build assets: %v", err)
+				return
+			}
+			mgr.storage.ReportBuildAssets(build, assets...)
```
Now that the dashapi.Build object includes Assets field, why don't we upload assets first and then in the Build object? This would exclude the whole issue of having inconsistent builds w/o assets.
Currently if syz-ci crashes, exits, restarted, VM crashes before the assets are uploaded, we won't have assets for this build permanently, right? We re-upload only in debug mode, but I assume it won't be generally enabled.

a-nogikh replied:
Sounds reasonable, updated the PR.

I was worried about the time to upload the assets, but it turns out it's quite fast anyway and actually negligible if compared to the time needed to rebuild a kernel.

a-nogikh commented on syz-ci/syz-ci.go:
```diff
@@ -271,30 +270,36 @@ func main() {
+		log.Logf(0, "deprecating assets")
+		wg.Add(1)
 		err := storage.DeprecateAssets()
+		wg.Done()
```
This was done to address this comment: https://github.com/google/syzkaller/pull/3223#discussion_r923354365

Could you please elaborate, what do you mean by goroutine owning a wait group count? It's a mechanism supposed to be used by multiple goroutines, that's why I get confused by it.

a-nogikh replied:
Discussed offline; fixed.

dvyukov commented on dashboard/dashapi/dashapi.go:
```diff
@@ -548,19 +559,21 @@ func (dash *Dashboard) UploadManagerStats(req *ManagerStatsReq) error {
+// 2. dashboard periodically checks if the asset is still needed.
+// 3. syz-ci queries needed_assets to figure out which assets are still needed.
+// 4. Once an asset is not needed, syz-ci removes the corresponding file.
+type NewAsset struct {
```
Ah, I see.
I don't have strong preference. We can always rename this.

a-nogikh commented on dashboard/app/api.go:
```diff
@@ -419,10 +421,26 @@ func apiUploadBuild(c context.Context, ns string, r *http.Request, payload []byt
+	}
 	if build, err := loadBuild(c, ns, req.ID); err == nil {
+		if len(newAssets) > 0 {
+			updatedBuild, err := appendBuildAssets(c, ns, req.ID, newAssets)
```
As of now, it should be empty during restarts if debug is off and contain the same assets (but possibly with different download URLs *) if it's on. I just needed to somehow handle this corner case for the completeness of the API, so I invoke the same code that processes the `add_build_assets` call. 

*) I'm planning to address the issue of unnecessary re-uploading in an extra commit. I'll push it tomorrow.
